pkg/repository/selector_matcher.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 VMware
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package repository
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/vmware-tanzu/cartographer/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/cartographer/pkg/selector"
+)
+
+// * handle errors correctly (panic!)
+
+// Selector interface is an object we can find selectors on, and identify in errors
+// TODO: Refactor GetMatchLabels, GetMatchFields and GetMatchLabelExpressions as
+//  GetSelector() v1alpha1.Selector
+//  and make that return type:
+//  type Selector struct {
+//      metav1.LabelSelector
+//  	SelectorMatchFields []v1alpha1.FieldSelectorRequirement
+//  }
+
+type Selector interface {
+	GetMatchLabels() labels.Set
+	GetMatchFields() []v1alpha1.FieldSelectorRequirement
+	GetObjectKind() schema.ObjectKind
+	GetName() string
+	GetMatchExpressions() []metav1.LabelSelectorRequirement
+}
+
+type Selectable interface {
+	GetLabels() map[string]string
+}
+
+// BestSelectorMatch attempts at finding the selectors that best match their selectors
+// against the selectors.
+func BestSelectorMatch(selectable Selectable, blueprints []Selector) ([]Selector, error) {
+
+	if len(blueprints) == 0 {
+		return nil, nil
+	}
+
+	var matchingSelectors = map[int][]Selector{}
+	var highWaterMark = 0
+
+	for _, target := range blueprints {
+		size := 0
+		ls := &metav1.LabelSelector{
+			MatchLabels:      target.GetMatchLabels(),
+			MatchExpressions: target.GetMatchExpressions(),
+		}
+
+		// -- Labels
+		sel, err := metav1.LabelSelectorAsSelector(ls)
+		if err != nil {
+			// TODO deal with me
```

Comment:
should we deal with it before the final merge?

Reply:
we said no in the PR notes